File: java/easer-insights/src/main/java/io/github/matteobertozzi/easerinsights/metrics/collectors/TimeRangeCounter.java
/*
 * Licensed to the Apache Software Foundation (ASF) under one or more
 * contributor license agreements.  See the NOTICE file distributed with
 * this work for additional information regarding copyright ownership.
 * The ASF licenses this file to You under the Apache License, Version 2.0
 * (the "License"); you may not use this file except in compliance with
 * the License.  You may obtain a copy of the License at
 *
 *      http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

package io.github.matteobertozzi.easerinsights.metrics.collectors;

import java.util.concurrent.TimeUnit;

import io.github.matteobertozzi.easerinsights.metrics.MetricDatumCollector;
import io.github.matteobertozzi.easerinsights.metrics.MetricDefinition;
import io.github.matteobertozzi.easerinsights.util.DatumUnitConverter;

public abstract class TimeRangeCounter implements MetricDatumCollector {
  protected TimeRangeCounter() {
    // no-op
  }

  @Override
  public String type() {
    return "TIME_RANGE_COUNTER";
  }

  // ==========================================================================================
  public static TimeRangeCounter newSingleThreaded(final long maxInterval, final long window, final TimeUnit unit) {
    return new TimeRangeCounterImplSt(maxInterval, window, unit);
  }

  public static TimeRangeCounter newMultiThreaded(final long maxInterval, final long window, final TimeUnit unit) {
    return new TimeRangeCounterImplMt(maxInterval, window, unit);
  }

  // ==========================================================================================
  protected abstract long add(long timestamp, long delta);

  @Override
  public void update(final long timestamp, final long value) {
    add(timestamp, value);
  }

  // ==========================================================================================
  protected static final TimeRangeCounterSnapshot EMPTY_SNAPSHOT = new TimeRangeCounterSnapshot(0, 0, new long[0]);
  public record TimeRangeCounterSnapshot (long lastInterval, long window, long[] counters) implements MetricDataSnapshot {
    public long getFirstInterval() {
      return lastInterval - (counters.length * window);
    }

    public long getLastInterval() {
      return lastInterval + window;
    }

    @Override
    public StringBuilder addToHumanReport(final MetricDefinition metricDefinition, final StringBuilder report) {
      if (window == 0) return report.append("(no data)\n");

      final DatumUnitConverter unitConverter = DatumUnitConverter.humanConverter(metricDefinition.unit());

      report.append("window ").append(DatumUnitConverter.humanTimeMillis(window));
      report.append(" - ").append(DatumUnitConverter.humanDateFromEpochMillis(getFirstInterval()));
      report.append(" - [");
      for (int i = 0, n = counters.length; i < n; ++i) {
        if (i > 0) report.append(',');
        report.append(unitConverter.asHumanString(counters[i]));
      }
      report.append("] - ");
      report.append(DatumUnitConverter.humanDateFromEpochMillis(getLastInterval()));
      report.append('\n');
      return report;
    }
  }
}

File: python/easer_insights/metrics/time_range_counter.py
# Licensed to the Apache Software Foundation (ASF) under one
# or more contributor license agreements.  See the NOTICE file
# distributed with this work for additional information
# regarding copyright ownership.  The ASF licenses this file
# to you under the Apache License, Version 2.0 (the
# "License"); you may not use this file except in compliance
# with the License.  You may obtain a copy of the License at
#
#     http://www.apache.org/licenses/LICENSE-2.0
#
# Unless required by applicable law or agreed to in writing, software
# distributed under the License is distributed on an "AS IS" BASIS,
# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
# See the License for the specific language governing permissions and
# limitations under the License.

from dataclasses import dataclass
from math import ceil

from .time_range import TimeRange
from ..util.time_unit import TimeUnit
from ..util.time_util import current_epoch_millis


@dataclass
class TimeRangeCounterSnapshot:
    last_interval: int
    window: int
    data: list[int]


class TimeRangeCounter:
    def __init__(self, max_interval: int, window: int, unit: TimeUnit) -> None:
        self._time_range = TimeRange(unit.to_millis(window), 0)
        trc_count = ceil(unit.to_millis(max_interval) / self._time_range.window)
        self._counters = [0] * trc_count

    def add(self, timestamp: int, delta: int):
        self._time_range.update(timestamp, len(self._counters), self._reset_slots,
                                lambda idx: self._update_slot(idx, delta))

    def set(self, timestamp: int, value: int):
        self._time_range.update(timestamp, len(self._counters), self._reset_slots,
                                lambda idx: self._assign_slot(idx, value))

    def _update_slot(self, index: int, delta: int):
        print(index)
        self._counters[index] += delta

    def _assign_slot(self, index: int, value: int):
        self._counters[index] = value

    def _reset_slots(self, from_index: int, to_index: int):
        for i in range(to_index - from_index):
            self._counters[from_index + i] = 0

    def snapshot(self) -> TimeRangeCounterSnapshot:
        now = current_epoch_millis()
        counters_length = len(self._counters)
        self._time_range.update(now, counters_length, self._reset_slots)

        data_size = self._time_range.size(counters_length)
        data = []
        self._time_range.copy(counters_length, data_size,
                              lambda _, src_from_idx, src_to_idx: data.extend(self._counters[src_from_idx:src_to_idx]))
        return TimeRangeCounterSnapshot(self._time_range.last_interval, self._time_range.window, data)
